detox/src/devices/drivers/android/tools/FreeDeviceFinderBase.test.js

```diff
@@ -0,0 +1,87 @@
+const FreeDeviceFinderBase = require('./FreeDeviceFinderBase');
+
+describe('devices lookup helper', () => {
+  const mockAdb = { devices: jest.fn() };
+
+  let mockDeviceRegistry;
+  let mockIsDeviceMatching;
+  let uut;
+  beforeEach(() => {
+    mockDeviceRegistry = { isDeviceBusy: jest.fn() };
+    mockIsDeviceMatching = jest.fn().mockResolvedValue(false);
+
+    const FreeDeviceFinderImpl = class extends FreeDeviceFinderBase {
+      constructor(adb, deviceRegistry) {
+        super(adb, deviceRegistry);
+        this.isDeviceMatching = mockIsDeviceMatching;
+      }
+    }
+    uut = new FreeDeviceFinderImpl(mockAdb, mockDeviceRegistry);
+  });
+
+  it('should return null when there are no devices', async () => {
+    mockNoAdbDevices();
+    mockAllDevicesFree();
+    mockAnyDeviceMatches();
+    expect(await uut.findFreeDevice()).toEqual(null);
+  });
+
+  it('should return null when all devices busy', async () => {
+    mockOneAdbDevice();
+    mockAllDevicesBusy();
+    mockAnyDeviceMatches();
+    expect(await uut.findFreeDevice()).toEqual(null);
+  });
+
+  it('should return null when no device matches', async () => {
+    const deviceQuery = 'mockDeviceQuery';
+    const candidate = createDevice();
+    mockAdbDevices([candidate]);
+    mockAllDevicesFree();
+    mockNoDeviceMatches();
+    expect(await uut.findFreeDevice(deviceQuery)).toEqual(null);
+    expect(mockIsDeviceMatching).toHaveBeenCalledWith(candidate, deviceQuery);
+  });
+
+  it('should return first device that matches', async () => {
+    const deviceQuery = 'mockDeviceQuery';
+    const candidate = createDevice('device2');
+    mockAdbDevices([
+      createDevice('device1'),
+      candidate,
+      createDevice('device3'),
+    ]);
+    mockAllDevicesFree();
+    mockNthDeviceMatches(2);
+    expect(await uut.findFreeDevice(deviceQuery)).toEqual(candidate.adbName);
+    expect(mockIsDeviceMatching).toHaveBeenCalledWith(candidate, deviceQuery);
+  });
+
+  it('should throw when `isDeviceMatching` is not overridden in a sub-class', async () => {
```

Comment:
Let's make it the first test in the suite.

Also, you can use `uut.findFreeDevice.restoreOriginalImplementation()` or something like that in Jest. No need to recreate it, I think. Look in the docs, it should be there.

Reply:
> Let's make it the first test in the suite.

Done

> Also, you can use uut.findFreeDevice.restoreOriginalImplementation() or something like that in Jest. No need to recreate it, I think. Look in the docs, it should be there.

I simplified the mocking logic. However, I left the `uut` instantiation in `beforeEach` because other unit tests follow same approach (15 test suits, e.g. _src/devices/drivers/android/AndroidDriver.test.js_).